File: badUser.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> a;
vector<string> b;
vector<string> ans;
int check[8];
int answer;
set<vector<string>> s_check;

int judge(vector<string> k) {
    int flag = 0;
    for (int i = 0; i < k.size(); i++) {
        if (k[i].size() != b[i].size()) {
            flag = 1;
            break;
        }
        for (int j = 0; j < k[i].size(); j++) {
            if (!(k[i][j] == b[i][j] || b[i][j] == '*')) {
                flag = 1;
                break;
            }
        }
        if (flag == 1) break;
    }
    if (flag == 0) {
        sort(k.begin(), k.end());
        s_check.insert(k);
        return 1;
    } else
        return 0;
}

void foo() {
    if (ans.size() == b.size()) {
        vector<string> tmp = ans;
        sort(tmp.begin(), tmp.end());
        if (s_check.find(tmp) == s_check.end() && judge(ans)) {
            answer++;
        }
        return;
    }
    for (int i = 0; i < a.size(); i++) {
        if (check[i] == 0) {
            check[i] = 1;
            ans.push_back(a[i]);
            foo();
            ans.pop_back();
            check[i] = 0;
        }
    }
}

int solution(vector<string> user_id, vector<string> banned_id) {
    a = user_id;
    b = banned_id;
    foo();
    return answer;
}

File: badUser.js
let n, m;
let bt = [];
let check = [];
let user, banned;
let ma = {};
let ans = 0;
function foo() {
    if (bt.length === m) {
        let tmp = [...bt];
        tmp.sort();
        let str = "";
        for (let i = 0; i < m; i++) {
            str += tmp[i];
        }
        if (ma[str] === undefined) {
            let flag = 0;
            for (let i = 0; i < m; i++) {
                if (user[bt[i]].length !== banned[i].length) {
                    flag = 1;
                    break;
                }
                for (let j = 0; j < m; j++) {
                    if (
                        !(
                            banned[i][j] === "*" ||
                            user[bt[i]][j] === banned[i][j]
                        )
                    ) {
                        flag = 1;
                        break;
                    }
                }
            }
            if (flag === 0) {
                ma[str] = 1;
                ans++;
            }
        }
        return;
    }
    for (let i = 0; i < n; i++) {
        if (check[i] === undefined) {
            check[i] = 1;
            bt.push(i);
            foo();
            bt.pop();
            check[i] = undefined;
        }
    }
}

function solution(user_id, banned_id) {
    n = user_id.length;
    m = banned_id.length;
    user = user_id;
    banned = banned_id;
    foo();
    return ans;
}
